lib/services/access_request.go

```diff
@@ -150,6 +150,10 @@ type DynamicAccess interface {
 	GetAccessRequests(ctx context.Context, filter AccessRequestFilter) ([]AccessRequest, error)
 	// DeleteAccessRequest deletes an access request.
 	DeleteAccessRequest(ctx context.Context, reqID string) error
+	// GetPluginData loads all plugin data matching the supplied filter.
+	GetPluginData(ctx context.Context, filter PluginDataFilter) ([]PluginData, error)
```

Comment:
This method attempts to merge two functions: `GetPluginData` and `GetPluginDatas` it seems like, what is inconsistent with the rest of the API. Is there a reason for that?

Reply:
This is consistent with how `GetAccessRequests` works.  There is no `GetAccessRequest` method; if you want to get a single resource, you just supply the resource name as part of the filter.

In terms of why this model originally arose, consistency with the new [watcher filters](https://github.com/gravitational/teleport/blob/798d2b552b74f9fec0f07cf0b9f47a01fd641a8a/lib/services/events.go#L51-L53) and optional parameters like `State` were a big factor.  It also played nicer with GRPC since a single message encapsulated all arguments.

Over the course of implementing this strategy for the `AccessRequest` resource it became obvious that creating a special method for getting a *single* access request would add a lot of additional code (new ephemeral GRPC messages in `auth.proto`, new methods for all of the different auth clients, new tests, and and so on) without any real benefit (a filter could already match a single resource).  I decided that a `GetAccessRequest` method was basically consistency hobgoblins at that point.

When it came time to implement the `PluginData` API, there were enough free parameters that the `Filter` strategy seemed appropriate here as well.  Like with the `AccessRequest` API, once a `Filter` based method exists, I just don't see the need for a single-resource getter.